cmd/releaseagent/internal/releasesteps/releasesteps.go

```diff
@@ -0,0 +1,537 @@
+	PollImagesCommit(ctx context.Context, versions []string, secret *Secret) (string, error)
+	CheckLatestMARGoVersion(ctx context.Context, versions []string) error
+	CreateAnnouncementBlogFile(ctx context.Context, versions []string, user string, security bool, secret *Secret) error
+}
+
+// CreateStepGraph creates the steps for a release of one or more versions of Microsoft Go. The
+// returned step graph is not running.
+//
+// If rs is nil, creates a new empty state that indicates no release work has been done yet.
+// Otherwise, rs is used to resume an existing release. Returns rs or the new State so it can be
+// used to resume a future release.
+//
+// While any step is running, it may modify State, so it is not safe to access the returned State.
+// When all steps are complete (success or fail), State can then be safely used.
+//
+// Implementation note: this function should only contain coordination code (moving inputs/outputs
+// between steps through the State and synchronizing). All work involving external resources should
+// be done by calling methods on the ServiceBundle.
+func CreateStepGraph(ri *Input, secret *Secret, rs *State, sb ServiceBundle) ([]*coordinator.Step, *State, error) {
+	if ri == nil || len(ri.Versions) == 0 {
+		return nil, nil, fmt.Errorf("no versions to release")
+	}
+
+	// Don't use simple "err" variable name here to avoid having "err" in scope during step
+	// creation. It is easy to accidentally capture it while writing new steps, and that results in
+	// a data race.
+	riChecksum, checksumErr := ri.checksum()
+	if checksumErr != nil {
+		return nil, nil, fmt.Errorf("failed to checksum release input: %v", checksumErr)
+	}
+
+	// Either create a new state or validate the existing one's checksum.
+	if rs == nil {
+		rs = &State{
+			InputChecksum: riChecksum,
+		}
+	} else if riChecksum != rs.InputChecksum {
+		return nil, nil, fmt.Errorf(
+			"release input doesn't match initial input: expected checksum %v (from state), got %v (by calculation)",
+			rs.InputChecksum, riChecksum)
+	}
+
+	// Ensure state is initialized.
+	if rs.Versions == nil {
+		rs.Versions = make(map[string]*VersionState)
+	}
+	for _, version := range ri.Versions {
+		if _, ok := rs.Versions[version]; !ok {
+			rs.Versions[version] = &VersionState{}
+		}
+	}
+
+	createStatusReportIssue := coordinator.NewRootStep(
+		"Create release day issue", coordinator.NoTimeout,
+		func(ctx context.Context) error {
+			if rs.Day.ReleaseIssue != 0 {
+				return nil
+			}
+			var err error
+			rs.Day.ReleaseIssue, err = sb.CreateReleaseDayTrackingIssue(
+				ctx, ri.TargetRepo, ri.RunnerGitHubUser, ri.Versions, secret)
+			return err
+		},
+	)
+
+	var versionCompleteSteps []*coordinator.Step
+	var versionSpecificPublishSteps []*coordinator.Step
+
+	for _, version := range ri.Versions {
+		vs := rs.Versions[version]
+		name := func(n string) string {
+			return fmt.Sprintf("%s, %s", n, version)
+		}
+
+		syncUpdate := coordinator.NewStep(
```

Comment:
Meta comment: it is difficult for me to see if a step timeout is just a generic "slow" vs "long" timeout, or if has hand-crafted for the given step. I would suggest defining some generic timeouts (i.e. fast=15min and slow=3h) and use hard-coded timeouts for those tasks that can't fit into the generic ones (adding the appropriate explanation).

Reply:
Good idea. For a handful of these I was weirdly specific, too, and they should really be generic until there's good reason.  (E.g. I guessed that because creating a GitHub release involves uploading assets, we should arbitrarily let it take a handful more minutes? But checking the most recent release, tagging took <1 second and creating a release took 5 seconds. 😄)

With the more easily extensible logic here, we can also do nifty things like keep track of each individual timeout and raise an alert if there's less than 1/3 of the timeout time remaining, and tune it based on real trends.